### ISSUE TITLE
Omit deriving Generic on LicenseId

### ISSUE THREAD

**Issue by ndmitchell** [haskell/cabal#5893]
Compiling the LicenseId file on my Windows GHC 6.8.3 takes ~30s (measured while running Hadrian). Experimenting with a cut down version, it turned out that most of the time was due to the Generic derivation on LicenseId. That generic derivation is used to provide a Binary instance, but can be replaced trivially by:

```
instance Binary LicenseId where
    put = put . fromEnum
    get = toEnum <$> get
```

In my measurements this reduces the compilation time significantly, from 9s to 3s at `-O0` and 22s to 8s on my cutdown file. That seems a worthwhile speedup.

Given this is an exported type, it's technically a user visible API change, hence raising a bug rather than a PR.

**Comment by 23Skidoo** [comment]
I believe it should be fine to apply this change to `master`.
6.8.3 is quite old though, how do more recent versions fare?

**Comment by ndmitchell** [comment]
Sorry, I meant 8.6.3 - the latest released one 🤦‍♂️

**Comment by 23Skidoo** [comment]
OK, let's do it.

**Comment by ndmitchell** [comment]
Hmm, actually, it's going to change the Binary instance for LicenseId, so things saved with the old won't still work. Is that an issue?

**Comment by 23Skidoo** [comment]
It should be fine, since we don't support reading old local build infos. May be worth bumping the `master` version to 2.5.1 so that people using `master` snapshots don't get weird errors.

**Comment by hvr** [comment]
I'm strongly against this, as this would mean that we'd end up with orphan instances somewhere, cause `Generic` instances are useful especially for non-trivial types such as LicenseId.

**Comment by 23Skidoo** [comment]
I don't understand the argument about orphan instances.

**Comment by harpocrates** [comment]
I think Herbert is suggesting that consumers of `Cabal` are likely to be interested in `Generic` instances of non-trivial types such as `LicenseId` and may end up declaring their own orphan instances of `Generic LicenseId`.

**Comment by 23Skidoo** [comment]
OK, makes sense.

**Comment by 23Skidoo** [comment]
For reference: [Neil's post about the problem](https://neilmitchell.blogspot.com/2019/02/quadratic-deriving-generic-compile-times.html); [GHC ticket](https://ghc.haskell.org/trac/ghc/ticket/5642).

**Comment by 23Skidoo** [comment]
Since there are strong objections, I'm closing this as wontfix.

**Comment by hsyl20** [comment]
Binary instance not relying on `Generics` has been implemented in #6001 / d8740cdd050cbc7e21b359524e82524055a1ea3a

**Comment by andreasabel** [comment]
Recent reinkarnations in:
- #8074 
- #8117
